{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fa3dbfd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import csv\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d4dadea",
   "metadata": {},
   "source": [
    "# Data Processing\n",
    "\n",
    "Read in and format the raw data from running CLEAN-Contact, CLEAN, and from grabbing data from BioCyc on Prochlorococcus marinus subsp. pastoris str. CCMP1986, referred to here as MED4. These changes will make them easier to compare and analyze later.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e8d998fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# relative paths for any OS\n",
    "# clean-contact results\n",
    "cc_path = os.path.join('..', 'raw-data', 'clean-contact')\n",
    "cc_maxsep_conf_path = os.path.join(cc_path, 'med4_uniprotkb_proteome_UP000001026_2024_01_13_only_ids_in_headers_maxsep_confidence.csv')\n",
    "cc_maxsep_path = os.path.join(cc_path, 'med4_uniprotkb_proteome_UP000001026_2024_01_13_only_ids_in_headers_maxsep.csv')\n",
    "cc_pvalue_conf_path = os.path.join(cc_path, 'med4_uniprotkb_proteome_UP000001026_2024_01_13_only_ids_in_headers_pvalue_confidence.csv')\n",
    "cc_pvalue_path = os.path.join(cc_path, 'med4_uniprotkb_proteome_UP000001026_2024_01_13_only_ids_in_headers_pvalue.csv')\n",
    "\n",
    "# clean results\n",
    "clean_path = os.path.join('..', 'raw-data', 'clean')\n",
    "clean_maxsep_conf_path = os.path.join(clean_path, 'med4_uniprotkb_proteome_UP000001026_2024_01_13_maxsep.csv')\n",
    "clean_pvalue_path = os.path.join(clean_path, 'med4_uniprotkb_proteome_UP000001026_2024_01_13_pvalue.csv')\n",
    "\n",
    "# med4 annotations from BioCyc: https://cyanocyc.org/group?id=biocyc18-15682-3934545975\n",
    "biocyc_path = os.path.join('..', 'raw-data', 'biocyc', 'Genes-of-a-reaction-from-All-instances-of-Reactions-in-Prochlorococcus-marinus-pastoris-CCMP1986.txt')\n",
    "\n",
    "# read in original med4 fasta to get the header line annotations\n",
    "med4_fasta_path = os.path.join('..', '..', 'computation', 'clean-contact', 'inputs', 'med4_uniprotkb_proteome_UP000001026_2024_01_13.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8c7b77b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_csv_row(row):\n",
    "    ''' \n",
    "    Helper function to read in csv data in correct format.\n",
    "    '''\n",
    "    # split the row by commas\n",
    "    values = row.split(',')\n",
    "    \n",
    "    # first value goes to the first column\n",
    "    first_value = values[0]\n",
    "    \n",
    "    # remaining values go to the second column as a list\n",
    "    remaining_values = values[1:]\n",
    "    \n",
    "    return [first_value, remaining_values]\n",
    "\n",
    "def process_results(data_path, my_cols):\n",
    "    '''\n",
    "    Read in csv results and give column names.\n",
    "    '''\n",
    "    # assumes no header lines in this input csv\n",
    "    processed_data = []\n",
    "    with open(data_path, 'r') as csv_file:\n",
    "        for line in csv_file:\n",
    "            processed_data.append(process_csv_row(line.strip()))\n",
    "    processed_df = pd.DataFrame(processed_data, columns=my_cols)\n",
    "    return processed_df\n",
    "\n",
    "def remove_extra_id_format(s):\n",
    "    '''\n",
    "    From 'tr|Q7TU21|Q7TU21_PROMP' to 'Q7TU21'.\n",
    "    '''\n",
    "    parts = s.split('|')\n",
    "    return f'{parts[1]}'\n",
    "\n",
    "def format_ec(ec_number):\n",
    "    ''' \n",
    "    Edit the BioCyc formatting for the EC number to match the format used by CLEAN and CLEAN-Contact.\n",
    "    Instead of 'EC-1.2.1.38 // EC-1.2.1.106', use 'EC:1.2.1.38, EC:1.2.1.106'.\n",
    "    '''\n",
    "    if pd.isna(ec_number):\n",
    "        return ec_number\n",
    "    ec_number = ec_number.replace(' // ', ', ')\n",
    "    entries = ec_number.split(', ') \n",
    "    clean_entries = [entry.replace('EC-', 'EC:') for entry in entries]\n",
    "    return ', '.join(clean_entries)\n",
    "\n",
    "def extract_ec_values(s):\n",
    "    ''' \n",
    "    From the full format of the CLEAN or CLEAN-Contact results column to just the EC numbers. \n",
    "    before: '[EC:#.#.#.#/#_#]'\n",
    "    after: 'EC:#.#.#.#'\n",
    "    '''\n",
    "    # regular expression handles n values in the last position of the ec number\n",
    "    ec_values = re.findall(r'EC:\\d+\\.\\d+\\.\\d+(?:\\.\\d+|\\.n\\d*)?', str(s))\n",
    "    # return results as comma separated string, not actual list datatype\n",
    "    return ', '.join(ec_values) \n",
    "\n",
    "def parse_fasta_ids(fasta_file):\n",
    "    '''\n",
    "    Parses a fasta file with UniProt ids and returns a dictionary of annotations and a dictionary with protein sequences \n",
    "    both with their corresponding UniProt ids.\n",
    "    '''\n",
    "    id_to_annotation = {}\n",
    "    id_to_sequence = {}\n",
    "    \n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        header = record.description\n",
    "        id_part = header.split('|')[1]\n",
    "        annotation_part = ' '.join(header.split(' ')[1:])\n",
    "        sequence = str(record.seq)\n",
    "        \n",
    "        id_to_annotation[id_part] = annotation_part\n",
    "        id_to_sequence[id_part] = sequence\n",
    "        \n",
    "    return id_to_annotation, id_to_sequence\n",
    "\n",
    "def update_dataframe_with_annotations(fasta_file, dataframe):\n",
    "    '''\n",
    "    Add annotation info and protein sequences from fasta file to results df based on common UniProt ids.\n",
    "    '''\n",
    "    id_to_annotation, id_to_sequence = parse_fasta_ids(fasta_file)\n",
    "    dataframe['Annotation'] = dataframe['UniProt'].map(id_to_annotation)\n",
    "    dataframe['Sequence'] = dataframe['UniProt'].map(id_to_sequence)\n",
    "    \n",
    "    return dataframe\n",
    "\n",
    "def extract_conf_no_std(string):\n",
    "    '''\n",
    "    Extract the confidence score part of the results, before the underscore.\n",
    "    '''\n",
    "    match = re.search(r'/([\\d.]+)_', string)\n",
    "    if match:\n",
    "        return match.group(1)\n",
    "    return None\n",
    "\n",
    "def extract_conf_only(lst):\n",
    "    '''\n",
    "    From the full format of CLEAN-Contact results column to just the confidence score, without the standard deviation range.\n",
    "    '''\n",
    "    return ', '.join(filter(None, [extract_conf_no_std(s) for s in lst]))\n",
    "\n",
    "\n",
    "def extract_std_range(string):\n",
    "    '''\n",
    "    Grab everything after the underscore.\n",
    "    '''\n",
    "    match = re.search(r'_([\\d.]+)', string)\n",
    "    if match:\n",
    "        return match.group(1)\n",
    "    return None\n",
    "\n",
    "def extract_std_only(lst):\n",
    "    '''\n",
    "    Apply the extraction of standard deviation range of the confidence scores on lists of strings.\n",
    "    '''\n",
    "    return ', '.join(filter(None, [extract_std_range(s) for s in lst]))\n",
    "\n",
    "\n",
    "def extract_value(string):\n",
    "    '''\n",
    "    Grab the value after the EC number prediction, so anything after /.\n",
    "    '''\n",
    "    match = re.search(r'/([\\d.]+)', string)\n",
    "    if match:\n",
    "        return match.group(1)\n",
    "    return None\n",
    "\n",
    "\n",
    "def extract_from_list(lst):\n",
    "    '''\n",
    "    Apply the extraction of values after the EC number on lists of strings.\n",
    "    '''\n",
    "    return ', '.join(filter(None, [extract_value(s) for s in lst]))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d0437d3",
   "metadata": {},
   "source": [
    "## CLEAN-Contact \n",
    "When running CLEAN-Contact using either inference method with the GMM, the GMM confidence score results are in addition to the results without running the GMM. For the case study, each inference method was run once with the GMM. This is why we start with 2 datasets for maxsep and 2 datasets for pvalue; each inference method has results with and without the GMM confidence scores. Also note that CLEAN-Contact's GMM results include the confidence score as well as the standard deviation for the confidence score.\n",
    "\n",
    "\n",
    "Maxsep\n",
    "* results from using the maxsep inference method, with the GMM to get confidence scores\n",
    "* results from using the maxsep inference method, without the GMM\n",
    "* combine the maxsep results and share in this directory as **cc_maxsep_results_full.csv**\n",
    "    * 1939 rows of data\n",
    "    * 7 columns\n",
    "        * 'UniProt': UniProt id\n",
    "        * 'Annotation': header annotations from the fasta file\n",
    "        * 'Sequence': protein sequences from the fasta file\n",
    "        * 'CC maxsep EC': CLEAN-Contact's EC predictions from using the maxsep inference method\n",
    "        * 'CC maxsep conf': CLEAN-Contact's confidence scores from using the maxsep inference method with the GMM\n",
    "        * 'CC maxsep conf SD': CLEAN-Contact's confidence score standard deviation from using the maxsep inference method with GMM\n",
    "        * 'CC maxsep cluster dist': CLEAN-Contact's EC prediction's distance from the cluster center from using the maxsep inference method\n",
    "\n",
    "Pvalue\n",
    "* results from using the pvalue inference method, with the GMM to get confidence scores\n",
    "* results from using the pvalue inference methods, without the GMM\n",
    "* combine the pvalue results and share in this directory as **cc_pvalue_results_full.csv**\n",
    "    * 1939 rows of data\n",
    "    * 7 columns\n",
    "        * 'UniProt': UniProt id\n",
    "        * 'Annotation': header annotations from the fasta file\n",
    "        * 'Sequence': protein sequences from the fasta file\n",
    "        * 'CC pvalue EC': CLEAN-Contact's EC predictions from using the pvalue inference method\n",
    "        * 'CC pvalue conf': CLEAN-Contact's confidence scores from using the pvalue inference method with the GMM\n",
    "        * 'CC pvalue conf SD': CLEAN-Contact's confidence score standard deviation from using the pvalue inference method with GMM\n",
    "        * 'CC pvalue cluster dist': CLEAN-Contact's EC prediction's distance from the cluster center from using the pvalue inference method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2b0767b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>CC maxsep conf full</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>CC maxsep conf EC</th>\n",
       "      <th>CC maxsep conf</th>\n",
       "      <th>CC maxsep conf SD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q7TU21</td>\n",
       "      <td>[EC:1.2.1.11/0.9863_0.0041]</td>\n",
       "      <td>Aspartate-semialdehyde dehydrogenase OS=Prochl...</td>\n",
       "      <td>MRNSPFLPNRPLKVAVLGSSGAVGSELLKILEERDFPISELVLLSS...</td>\n",
       "      <td>EC:1.2.1.11</td>\n",
       "      <td>0.9863</td>\n",
       "      <td>0.0041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q7UZH7</td>\n",
       "      <td>[EC:6.3.4.2/0.9961_0.0017]</td>\n",
       "      <td>CTP synthase OS=Prochlorococcus marinus subsp....</td>\n",
       "      <td>MSKFVFVTGGVVSSIGKGIVAASLGRLLKSRGYSVSILKLDPYLNV...</td>\n",
       "      <td>EC:6.3.4.2</td>\n",
       "      <td>0.9961</td>\n",
       "      <td>0.0017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q7UZR6</td>\n",
       "      <td>[EC:6.3.2.1/0.8120_0.0551]</td>\n",
       "      <td>Bifunctional pantoate ligase/cytidylate kinase...</td>\n",
       "      <td>MNKIIIRKTEDLKEWRRNLKCDINFIPTMGNLHDGHQKLISTAQSS...</td>\n",
       "      <td>EC:6.3.2.1</td>\n",
       "      <td>0.8120</td>\n",
       "      <td>0.0551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q7V010</td>\n",
       "      <td>[EC:2.1.1.192/0.9923_0.0025]</td>\n",
       "      <td>Probable dual-specificity RNA methyltransferas...</td>\n",
       "      <td>MKNLLGCSVKDLEKIALNYGQAAFRGRQIYNWLYNYKNRSKSIDEI...</td>\n",
       "      <td>EC:2.1.1.192</td>\n",
       "      <td>0.9923</td>\n",
       "      <td>0.0025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q7V0C4</td>\n",
       "      <td>[EC:2.7.11.1/0.0911_0.0452]</td>\n",
       "      <td>Circadian clock oscillator protein KaiC OS=Pro...</td>\n",
       "      <td>MKDKKISKSIKMQVQKIPTGIEGFDDVCRGGLPAARSTLVSGTSGT...</td>\n",
       "      <td>EC:2.7.11.1</td>\n",
       "      <td>0.0911</td>\n",
       "      <td>0.0452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1934</th>\n",
       "      <td>Q7V3R0</td>\n",
       "      <td>[EC:4.2.3.187/0.0014_0.0012]</td>\n",
       "      <td>Antitermination protein NusB OS=Prochlorococcu...</td>\n",
       "      <td>MMHNRSLSRELSLLSLGLIKDTADLELNKIQIDEIFESALDSLINH...</td>\n",
       "      <td>EC:4.2.3.187</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1935</th>\n",
       "      <td>Q7V3R1</td>\n",
       "      <td>[EC:2.7.1.66/0.0014_0.0012]</td>\n",
       "      <td>Transporter OS=Prochlorococcus marinus subsp. ...</td>\n",
       "      <td>MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...</td>\n",
       "      <td>EC:2.7.1.66</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1936</th>\n",
       "      <td>Q7V3R2</td>\n",
       "      <td>[EC:1.17.99.6/0.9914_0.0028]</td>\n",
       "      <td>4Fe-4S ferredoxin-type domain-containing prote...</td>\n",
       "      <td>MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...</td>\n",
       "      <td>EC:1.17.99.6</td>\n",
       "      <td>0.9914</td>\n",
       "      <td>0.0028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937</th>\n",
       "      <td>Q7V3R3</td>\n",
       "      <td>[EC:3.5.2.6/0.0787_0.0403]</td>\n",
       "      <td>Uncharacterized protein OS=Prochlorococcus mar...</td>\n",
       "      <td>MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...</td>\n",
       "      <td>EC:3.5.2.6</td>\n",
       "      <td>0.0787</td>\n",
       "      <td>0.0403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1938</th>\n",
       "      <td>Q7V3R6</td>\n",
       "      <td>[EC:6.2.1.57/0.0001_0.0001]</td>\n",
       "      <td>PRC-barrel domain-containing protein OS=Prochl...</td>\n",
       "      <td>MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...</td>\n",
       "      <td>EC:6.2.1.57</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1939 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     UniProt           CC maxsep conf full  \\\n",
       "0     Q7TU21   [EC:1.2.1.11/0.9863_0.0041]   \n",
       "1     Q7UZH7    [EC:6.3.4.2/0.9961_0.0017]   \n",
       "2     Q7UZR6    [EC:6.3.2.1/0.8120_0.0551]   \n",
       "3     Q7V010  [EC:2.1.1.192/0.9923_0.0025]   \n",
       "4     Q7V0C4   [EC:2.7.11.1/0.0911_0.0452]   \n",
       "...      ...                           ...   \n",
       "1934  Q7V3R0  [EC:4.2.3.187/0.0014_0.0012]   \n",
       "1935  Q7V3R1   [EC:2.7.1.66/0.0014_0.0012]   \n",
       "1936  Q7V3R2  [EC:1.17.99.6/0.9914_0.0028]   \n",
       "1937  Q7V3R3    [EC:3.5.2.6/0.0787_0.0403]   \n",
       "1938  Q7V3R6   [EC:6.2.1.57/0.0001_0.0001]   \n",
       "\n",
       "                                             Annotation  \\\n",
       "0     Aspartate-semialdehyde dehydrogenase OS=Prochl...   \n",
       "1     CTP synthase OS=Prochlorococcus marinus subsp....   \n",
       "2     Bifunctional pantoate ligase/cytidylate kinase...   \n",
       "3     Probable dual-specificity RNA methyltransferas...   \n",
       "4     Circadian clock oscillator protein KaiC OS=Pro...   \n",
       "...                                                 ...   \n",
       "1934  Antitermination protein NusB OS=Prochlorococcu...   \n",
       "1935  Transporter OS=Prochlorococcus marinus subsp. ...   \n",
       "1936  4Fe-4S ferredoxin-type domain-containing prote...   \n",
       "1937  Uncharacterized protein OS=Prochlorococcus mar...   \n",
       "1938  PRC-barrel domain-containing protein OS=Prochl...   \n",
       "\n",
       "                                               Sequence CC maxsep conf EC  \\\n",
       "0     MRNSPFLPNRPLKVAVLGSSGAVGSELLKILEERDFPISELVLLSS...       EC:1.2.1.11   \n",
       "1     MSKFVFVTGGVVSSIGKGIVAASLGRLLKSRGYSVSILKLDPYLNV...        EC:6.3.4.2   \n",
       "2     MNKIIIRKTEDLKEWRRNLKCDINFIPTMGNLHDGHQKLISTAQSS...        EC:6.3.2.1   \n",
       "3     MKNLLGCSVKDLEKIALNYGQAAFRGRQIYNWLYNYKNRSKSIDEI...      EC:2.1.1.192   \n",
       "4     MKDKKISKSIKMQVQKIPTGIEGFDDVCRGGLPAARSTLVSGTSGT...       EC:2.7.11.1   \n",
       "...                                                 ...               ...   \n",
       "1934  MMHNRSLSRELSLLSLGLIKDTADLELNKIQIDEIFESALDSLINH...      EC:4.2.3.187   \n",
       "1935  MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...       EC:2.7.1.66   \n",
       "1936  MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...      EC:1.17.99.6   \n",
       "1937  MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...        EC:3.5.2.6   \n",
       "1938  MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...       EC:6.2.1.57   \n",
       "\n",
       "     CC maxsep conf CC maxsep conf SD  \n",
       "0            0.9863            0.0041  \n",
       "1            0.9961            0.0017  \n",
       "2            0.8120            0.0551  \n",
       "3            0.9923            0.0025  \n",
       "4            0.0911            0.0452  \n",
       "...             ...               ...  \n",
       "1934         0.0014            0.0012  \n",
       "1935         0.0014            0.0012  \n",
       "1936         0.9914            0.0028  \n",
       "1937         0.0787            0.0403  \n",
       "1938         0.0001            0.0001  \n",
       "\n",
       "[1939 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CLEAN-Contact maxsep results, with confidence scores and SD from using the gmm\n",
    "cc_maxsep_conf = process_results(cc_maxsep_conf_path, ['UniProt', 'CC maxsep conf full'])\n",
    "cc_maxsep_conf = update_dataframe_with_annotations(med4_fasta_path, cc_maxsep_conf)\n",
    "cc_maxsep_conf['CC maxsep conf EC'] = cc_maxsep_conf['CC maxsep conf full'].apply(extract_ec_values)\n",
    "cc_maxsep_conf['CC maxsep conf'] = cc_maxsep_conf['CC maxsep conf full'].apply(extract_conf_only)\n",
    "cc_maxsep_conf['CC maxsep conf SD'] = cc_maxsep_conf['CC maxsep conf full'].apply(extract_std_only)\n",
    "cc_maxsep_conf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cfd1a9f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>CC maxsep full</th>\n",
       "      <th>CC maxsep EC</th>\n",
       "      <th>CC maxsep cluster dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q7TU21</td>\n",
       "      <td>[EC:1.2.1.11/4.3332]</td>\n",
       "      <td>EC:1.2.1.11</td>\n",
       "      <td>4.3332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q7UZH7</td>\n",
       "      <td>[EC:6.3.4.2/2.8527]</td>\n",
       "      <td>EC:6.3.4.2</td>\n",
       "      <td>2.8527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q7UZR6</td>\n",
       "      <td>[EC:6.3.2.1/5.8351]</td>\n",
       "      <td>EC:6.3.2.1</td>\n",
       "      <td>5.8351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q7V010</td>\n",
       "      <td>[EC:2.1.1.192/3.8433]</td>\n",
       "      <td>EC:2.1.1.192</td>\n",
       "      <td>3.8433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q7V0C4</td>\n",
       "      <td>[EC:2.7.11.1/7.2259]</td>\n",
       "      <td>EC:2.7.11.1</td>\n",
       "      <td>7.2259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1934</th>\n",
       "      <td>Q7V3R0</td>\n",
       "      <td>[EC:4.2.3.187/8.4480]</td>\n",
       "      <td>EC:4.2.3.187</td>\n",
       "      <td>8.4480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1935</th>\n",
       "      <td>Q7V3R1</td>\n",
       "      <td>[EC:2.7.1.66/8.4444]</td>\n",
       "      <td>EC:2.7.1.66</td>\n",
       "      <td>8.4444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1936</th>\n",
       "      <td>Q7V3R2</td>\n",
       "      <td>[EC:1.17.99.6/3.9525]</td>\n",
       "      <td>EC:1.17.99.6</td>\n",
       "      <td>3.9525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937</th>\n",
       "      <td>Q7V3R3</td>\n",
       "      <td>[EC:3.5.2.6/7.2768]</td>\n",
       "      <td>EC:3.5.2.6</td>\n",
       "      <td>7.2768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1938</th>\n",
       "      <td>Q7V3R6</td>\n",
       "      <td>[EC:6.2.1.57/9.1441]</td>\n",
       "      <td>EC:6.2.1.57</td>\n",
       "      <td>9.1441</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1939 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     UniProt         CC maxsep full  CC maxsep EC CC maxsep cluster dist\n",
       "0     Q7TU21   [EC:1.2.1.11/4.3332]   EC:1.2.1.11                 4.3332\n",
       "1     Q7UZH7    [EC:6.3.4.2/2.8527]    EC:6.3.4.2                 2.8527\n",
       "2     Q7UZR6    [EC:6.3.2.1/5.8351]    EC:6.3.2.1                 5.8351\n",
       "3     Q7V010  [EC:2.1.1.192/3.8433]  EC:2.1.1.192                 3.8433\n",
       "4     Q7V0C4   [EC:2.7.11.1/7.2259]   EC:2.7.11.1                 7.2259\n",
       "...      ...                    ...           ...                    ...\n",
       "1934  Q7V3R0  [EC:4.2.3.187/8.4480]  EC:4.2.3.187                 8.4480\n",
       "1935  Q7V3R1   [EC:2.7.1.66/8.4444]   EC:2.7.1.66                 8.4444\n",
       "1936  Q7V3R2  [EC:1.17.99.6/3.9525]  EC:1.17.99.6                 3.9525\n",
       "1937  Q7V3R3    [EC:3.5.2.6/7.2768]    EC:3.5.2.6                 7.2768\n",
       "1938  Q7V3R6   [EC:6.2.1.57/9.1441]   EC:6.2.1.57                 9.1441\n",
       "\n",
       "[1939 rows x 4 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CLEAN-Contact maxsep results, with distance from cluster center\n",
    "cc_maxsep = process_results(cc_maxsep_path, ['UniProt', 'CC maxsep full'])\n",
    "cc_maxsep['CC maxsep EC'] = cc_maxsep['CC maxsep full'].apply(extract_ec_values)\n",
    "cc_maxsep['CC maxsep cluster dist'] = cc_maxsep['CC maxsep full'].apply(extract_from_list)\n",
    "cc_maxsep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4df155c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The columns match. Proceeding to combine the DataFrames.\n"
     ]
    }
   ],
   "source": [
    "# combine and format the maxsep results if they have the same UniProt id and EC predictions\n",
    "if cc_maxsep_conf['CC maxsep conf EC'].equals(cc_maxsep['CC maxsep EC']):\n",
    "    print(\"The columns match. Proceeding to combine the DataFrames.\")\n",
    "    \n",
    "    # Combine the DataFrames on 'UniProt' and maxsep columns\n",
    "    cc_maxsep_results_full = pd.merge(cc_maxsep_conf, cc_maxsep, left_on=['UniProt', 'CC maxsep conf EC'], right_on=['UniProt', 'CC maxsep EC'])\n",
    "    \n",
    "    # Drop the redundant maxsep ec column \n",
    "    cc_maxsep_results_full.drop(columns=['CC maxsep conf EC'], inplace=True)\n",
    "    \n",
    "else:\n",
    "    print(\"The columns do not match.\")\n",
    "\n",
    "cc_maxsep_results_full.drop(columns=['CC maxsep conf full', 'CC maxsep full'], inplace=True)\n",
    "col_to_move = cc_maxsep_results_full.pop('CC maxsep EC')\n",
    "cc_maxsep_results_full.insert(3, 'CC maxsep EC', col_to_move)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cb89e83e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>CC maxsep EC</th>\n",
       "      <th>CC maxsep conf</th>\n",
       "      <th>CC maxsep conf SD</th>\n",
       "      <th>CC maxsep cluster dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q7TU21</td>\n",
       "      <td>Aspartate-semialdehyde dehydrogenase OS=Prochl...</td>\n",
       "      <td>MRNSPFLPNRPLKVAVLGSSGAVGSELLKILEERDFPISELVLLSS...</td>\n",
       "      <td>EC:1.2.1.11</td>\n",
       "      <td>0.9863</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>4.3332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q7UZH7</td>\n",
       "      <td>CTP synthase OS=Prochlorococcus marinus subsp....</td>\n",
       "      <td>MSKFVFVTGGVVSSIGKGIVAASLGRLLKSRGYSVSILKLDPYLNV...</td>\n",
       "      <td>EC:6.3.4.2</td>\n",
       "      <td>0.9961</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>2.8527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q7UZR6</td>\n",
       "      <td>Bifunctional pantoate ligase/cytidylate kinase...</td>\n",
       "      <td>MNKIIIRKTEDLKEWRRNLKCDINFIPTMGNLHDGHQKLISTAQSS...</td>\n",
       "      <td>EC:6.3.2.1</td>\n",
       "      <td>0.8120</td>\n",
       "      <td>0.0551</td>\n",
       "      <td>5.8351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q7V010</td>\n",
       "      <td>Probable dual-specificity RNA methyltransferas...</td>\n",
       "      <td>MKNLLGCSVKDLEKIALNYGQAAFRGRQIYNWLYNYKNRSKSIDEI...</td>\n",
       "      <td>EC:2.1.1.192</td>\n",
       "      <td>0.9923</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>3.8433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q7V0C4</td>\n",
       "      <td>Circadian clock oscillator protein KaiC OS=Pro...</td>\n",
       "      <td>MKDKKISKSIKMQVQKIPTGIEGFDDVCRGGLPAARSTLVSGTSGT...</td>\n",
       "      <td>EC:2.7.11.1</td>\n",
       "      <td>0.0911</td>\n",
       "      <td>0.0452</td>\n",
       "      <td>7.2259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1934</th>\n",
       "      <td>Q7V3R0</td>\n",
       "      <td>Antitermination protein NusB OS=Prochlorococcu...</td>\n",
       "      <td>MMHNRSLSRELSLLSLGLIKDTADLELNKIQIDEIFESALDSLINH...</td>\n",
       "      <td>EC:4.2.3.187</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0012</td>\n",
       "      <td>8.4480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1935</th>\n",
       "      <td>Q7V3R1</td>\n",
       "      <td>Transporter OS=Prochlorococcus marinus subsp. ...</td>\n",
       "      <td>MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...</td>\n",
       "      <td>EC:2.7.1.66</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0012</td>\n",
       "      <td>8.4444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1936</th>\n",
       "      <td>Q7V3R2</td>\n",
       "      <td>4Fe-4S ferredoxin-type domain-containing prote...</td>\n",
       "      <td>MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...</td>\n",
       "      <td>EC:1.17.99.6</td>\n",
       "      <td>0.9914</td>\n",
       "      <td>0.0028</td>\n",
       "      <td>3.9525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937</th>\n",
       "      <td>Q7V3R3</td>\n",
       "      <td>Uncharacterized protein OS=Prochlorococcus mar...</td>\n",
       "      <td>MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...</td>\n",
       "      <td>EC:3.5.2.6</td>\n",
       "      <td>0.0787</td>\n",
       "      <td>0.0403</td>\n",
       "      <td>7.2768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1938</th>\n",
       "      <td>Q7V3R6</td>\n",
       "      <td>PRC-barrel domain-containing protein OS=Prochl...</td>\n",
       "      <td>MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...</td>\n",
       "      <td>EC:6.2.1.57</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>9.1441</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1939 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     UniProt                                         Annotation  \\\n",
       "0     Q7TU21  Aspartate-semialdehyde dehydrogenase OS=Prochl...   \n",
       "1     Q7UZH7  CTP synthase OS=Prochlorococcus marinus subsp....   \n",
       "2     Q7UZR6  Bifunctional pantoate ligase/cytidylate kinase...   \n",
       "3     Q7V010  Probable dual-specificity RNA methyltransferas...   \n",
       "4     Q7V0C4  Circadian clock oscillator protein KaiC OS=Pro...   \n",
       "...      ...                                                ...   \n",
       "1934  Q7V3R0  Antitermination protein NusB OS=Prochlorococcu...   \n",
       "1935  Q7V3R1  Transporter OS=Prochlorococcus marinus subsp. ...   \n",
       "1936  Q7V3R2  4Fe-4S ferredoxin-type domain-containing prote...   \n",
       "1937  Q7V3R3  Uncharacterized protein OS=Prochlorococcus mar...   \n",
       "1938  Q7V3R6  PRC-barrel domain-containing protein OS=Prochl...   \n",
       "\n",
       "                                               Sequence  CC maxsep EC  \\\n",
       "0     MRNSPFLPNRPLKVAVLGSSGAVGSELLKILEERDFPISELVLLSS...   EC:1.2.1.11   \n",
       "1     MSKFVFVTGGVVSSIGKGIVAASLGRLLKSRGYSVSILKLDPYLNV...    EC:6.3.4.2   \n",
       "2     MNKIIIRKTEDLKEWRRNLKCDINFIPTMGNLHDGHQKLISTAQSS...    EC:6.3.2.1   \n",
       "3     MKNLLGCSVKDLEKIALNYGQAAFRGRQIYNWLYNYKNRSKSIDEI...  EC:2.1.1.192   \n",
       "4     MKDKKISKSIKMQVQKIPTGIEGFDDVCRGGLPAARSTLVSGTSGT...   EC:2.7.11.1   \n",
       "...                                                 ...           ...   \n",
       "1934  MMHNRSLSRELSLLSLGLIKDTADLELNKIQIDEIFESALDSLINH...  EC:4.2.3.187   \n",
       "1935  MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...   EC:2.7.1.66   \n",
       "1936  MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...  EC:1.17.99.6   \n",
       "1937  MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...    EC:3.5.2.6   \n",
       "1938  MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...   EC:6.2.1.57   \n",
       "\n",
       "     CC maxsep conf CC maxsep conf SD CC maxsep cluster dist  \n",
       "0            0.9863            0.0041                 4.3332  \n",
       "1            0.9961            0.0017                 2.8527  \n",
       "2            0.8120            0.0551                 5.8351  \n",
       "3            0.9923            0.0025                 3.8433  \n",
       "4            0.0911            0.0452                 7.2259  \n",
       "...             ...               ...                    ...  \n",
       "1934         0.0014            0.0012                 8.4480  \n",
       "1935         0.0014            0.0012                 8.4444  \n",
       "1936         0.9914            0.0028                 3.9525  \n",
       "1937         0.0787            0.0403                 7.2768  \n",
       "1938         0.0001            0.0001                 9.1441  \n",
       "\n",
       "[1939 rows x 7 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cc_maxsep_results_full.to_csv('cc_maxsep_results_full.csv', index=False)\n",
    "cc_maxsep_results_full"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c1137462",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>CC pvalue conf full</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>CC pvalue conf EC</th>\n",
       "      <th>CC pvalue conf</th>\n",
       "      <th>CC pvalue conf SD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q7TU21</td>\n",
       "      <td>[EC:1.2.1.11/0.9863_0.0041]</td>\n",
       "      <td>Aspartate-semialdehyde dehydrogenase OS=Prochl...</td>\n",
       "      <td>MRNSPFLPNRPLKVAVLGSSGAVGSELLKILEERDFPISELVLLSS...</td>\n",
       "      <td>EC:1.2.1.11</td>\n",
       "      <td>0.9863</td>\n",
       "      <td>0.0041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q7UZH7</td>\n",
       "      <td>[EC:6.3.4.2/0.9961_0.0017]</td>\n",
       "      <td>CTP synthase OS=Prochlorococcus marinus subsp....</td>\n",
       "      <td>MSKFVFVTGGVVSSIGKGIVAASLGRLLKSRGYSVSILKLDPYLNV...</td>\n",
       "      <td>EC:6.3.4.2</td>\n",
       "      <td>0.9961</td>\n",
       "      <td>0.0017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q7UZR6</td>\n",
       "      <td>[EC:6.3.2.1/0.8120_0.0551]</td>\n",
       "      <td>Bifunctional pantoate ligase/cytidylate kinase...</td>\n",
       "      <td>MNKIIIRKTEDLKEWRRNLKCDINFIPTMGNLHDGHQKLISTAQSS...</td>\n",
       "      <td>EC:6.3.2.1</td>\n",
       "      <td>0.8120</td>\n",
       "      <td>0.0551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q7V010</td>\n",
       "      <td>[EC:2.1.1.192/0.9923_0.0025]</td>\n",
       "      <td>Probable dual-specificity RNA methyltransferas...</td>\n",
       "      <td>MKNLLGCSVKDLEKIALNYGQAAFRGRQIYNWLYNYKNRSKSIDEI...</td>\n",
       "      <td>EC:2.1.1.192</td>\n",
       "      <td>0.9923</td>\n",
       "      <td>0.0025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q7V0C4</td>\n",
       "      <td>[EC:2.7.11.1/0.0911_0.0452]</td>\n",
       "      <td>Circadian clock oscillator protein KaiC OS=Pro...</td>\n",
       "      <td>MKDKKISKSIKMQVQKIPTGIEGFDDVCRGGLPAARSTLVSGTSGT...</td>\n",
       "      <td>EC:2.7.11.1</td>\n",
       "      <td>0.0911</td>\n",
       "      <td>0.0452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1934</th>\n",
       "      <td>Q7V3R0</td>\n",
       "      <td>[EC:4.2.3.187/0.0014_0.0012]</td>\n",
       "      <td>Antitermination protein NusB OS=Prochlorococcu...</td>\n",
       "      <td>MMHNRSLSRELSLLSLGLIKDTADLELNKIQIDEIFESALDSLINH...</td>\n",
       "      <td>EC:4.2.3.187</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1935</th>\n",
       "      <td>Q7V3R1</td>\n",
       "      <td>[EC:2.7.1.66/0.0014_0.0012]</td>\n",
       "      <td>Transporter OS=Prochlorococcus marinus subsp. ...</td>\n",
       "      <td>MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...</td>\n",
       "      <td>EC:2.7.1.66</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1936</th>\n",
       "      <td>Q7V3R2</td>\n",
       "      <td>[EC:1.17.99.6/0.9914_0.0028, EC:4.1.99.3/0.000...</td>\n",
       "      <td>4Fe-4S ferredoxin-type domain-containing prote...</td>\n",
       "      <td>MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...</td>\n",
       "      <td>EC:1.17.99.6, EC:4.1.99.3, EC:4.1.99.13</td>\n",
       "      <td>0.9914, 0.0000, 0.0000</td>\n",
       "      <td>0.0028, 0.0000, 0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937</th>\n",
       "      <td>Q7V3R3</td>\n",
       "      <td>[EC:3.5.2.6/0.0787_0.0403, EC:3.1.1.107/0.0034...</td>\n",
       "      <td>Uncharacterized protein OS=Prochlorococcus mar...</td>\n",
       "      <td>MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...</td>\n",
       "      <td>EC:3.5.2.6, EC:3.1.1.107</td>\n",
       "      <td>0.0787, 0.0034</td>\n",
       "      <td>0.0403, 0.0027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1938</th>\n",
       "      <td>Q7V3R6</td>\n",
       "      <td>[EC:6.2.1.57/0.0001_0.0001, EC:2.4.1.33/0.0000...</td>\n",
       "      <td>PRC-barrel domain-containing protein OS=Prochl...</td>\n",
       "      <td>MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...</td>\n",
       "      <td>EC:6.2.1.57, EC:2.4.1.33</td>\n",
       "      <td>0.0001, 0.0000</td>\n",
       "      <td>0.0001, 0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1939 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     UniProt                                CC pvalue conf full  \\\n",
       "0     Q7TU21                        [EC:1.2.1.11/0.9863_0.0041]   \n",
       "1     Q7UZH7                         [EC:6.3.4.2/0.9961_0.0017]   \n",
       "2     Q7UZR6                         [EC:6.3.2.1/0.8120_0.0551]   \n",
       "3     Q7V010                       [EC:2.1.1.192/0.9923_0.0025]   \n",
       "4     Q7V0C4                        [EC:2.7.11.1/0.0911_0.0452]   \n",
       "...      ...                                                ...   \n",
       "1934  Q7V3R0                       [EC:4.2.3.187/0.0014_0.0012]   \n",
       "1935  Q7V3R1                        [EC:2.7.1.66/0.0014_0.0012]   \n",
       "1936  Q7V3R2  [EC:1.17.99.6/0.9914_0.0028, EC:4.1.99.3/0.000...   \n",
       "1937  Q7V3R3  [EC:3.5.2.6/0.0787_0.0403, EC:3.1.1.107/0.0034...   \n",
       "1938  Q7V3R6  [EC:6.2.1.57/0.0001_0.0001, EC:2.4.1.33/0.0000...   \n",
       "\n",
       "                                             Annotation  \\\n",
       "0     Aspartate-semialdehyde dehydrogenase OS=Prochl...   \n",
       "1     CTP synthase OS=Prochlorococcus marinus subsp....   \n",
       "2     Bifunctional pantoate ligase/cytidylate kinase...   \n",
       "3     Probable dual-specificity RNA methyltransferas...   \n",
       "4     Circadian clock oscillator protein KaiC OS=Pro...   \n",
       "...                                                 ...   \n",
       "1934  Antitermination protein NusB OS=Prochlorococcu...   \n",
       "1935  Transporter OS=Prochlorococcus marinus subsp. ...   \n",
       "1936  4Fe-4S ferredoxin-type domain-containing prote...   \n",
       "1937  Uncharacterized protein OS=Prochlorococcus mar...   \n",
       "1938  PRC-barrel domain-containing protein OS=Prochl...   \n",
       "\n",
       "                                               Sequence  \\\n",
       "0     MRNSPFLPNRPLKVAVLGSSGAVGSELLKILEERDFPISELVLLSS...   \n",
       "1     MSKFVFVTGGVVSSIGKGIVAASLGRLLKSRGYSVSILKLDPYLNV...   \n",
       "2     MNKIIIRKTEDLKEWRRNLKCDINFIPTMGNLHDGHQKLISTAQSS...   \n",
       "3     MKNLLGCSVKDLEKIALNYGQAAFRGRQIYNWLYNYKNRSKSIDEI...   \n",
       "4     MKDKKISKSIKMQVQKIPTGIEGFDDVCRGGLPAARSTLVSGTSGT...   \n",
       "...                                                 ...   \n",
       "1934  MMHNRSLSRELSLLSLGLIKDTADLELNKIQIDEIFESALDSLINH...   \n",
       "1935  MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...   \n",
       "1936  MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...   \n",
       "1937  MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...   \n",
       "1938  MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...   \n",
       "\n",
       "                            CC pvalue conf EC          CC pvalue conf  \\\n",
       "0                                 EC:1.2.1.11                  0.9863   \n",
       "1                                  EC:6.3.4.2                  0.9961   \n",
       "2                                  EC:6.3.2.1                  0.8120   \n",
       "3                                EC:2.1.1.192                  0.9923   \n",
       "4                                 EC:2.7.11.1                  0.0911   \n",
       "...                                       ...                     ...   \n",
       "1934                             EC:4.2.3.187                  0.0014   \n",
       "1935                              EC:2.7.1.66                  0.0014   \n",
       "1936  EC:1.17.99.6, EC:4.1.99.3, EC:4.1.99.13  0.9914, 0.0000, 0.0000   \n",
       "1937                 EC:3.5.2.6, EC:3.1.1.107          0.0787, 0.0034   \n",
       "1938                 EC:6.2.1.57, EC:2.4.1.33          0.0001, 0.0000   \n",
       "\n",
       "           CC pvalue conf SD  \n",
       "0                     0.0041  \n",
       "1                     0.0017  \n",
       "2                     0.0551  \n",
       "3                     0.0025  \n",
       "4                     0.0452  \n",
       "...                      ...  \n",
       "1934                  0.0012  \n",
       "1935                  0.0012  \n",
       "1936  0.0028, 0.0000, 0.0000  \n",
       "1937          0.0403, 0.0027  \n",
       "1938          0.0001, 0.0000  \n",
       "\n",
       "[1939 rows x 7 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CLEAN-Contact pvalue results, with confidence scores and SD from using the gmm\n",
    "cc_pvalue_conf = process_results(cc_pvalue_conf_path, ['UniProt', 'CC pvalue conf full'])\n",
    "cc_pvalue_conf = update_dataframe_with_annotations(med4_fasta_path, cc_pvalue_conf)\n",
    "cc_pvalue_conf['CC pvalue conf EC'] = cc_pvalue_conf['CC pvalue conf full'].apply(extract_ec_values)\n",
    "cc_pvalue_conf['CC pvalue conf'] = cc_pvalue_conf['CC pvalue conf full'].apply(extract_conf_only)\n",
    "cc_pvalue_conf['CC pvalue conf SD'] = cc_pvalue_conf['CC pvalue conf full'].apply(extract_std_only)\n",
    "cc_pvalue_conf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e4a0ef7e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>CC pvalue full</th>\n",
       "      <th>CC pvalue EC</th>\n",
       "      <th>CC pvalue cluster dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q7TU21</td>\n",
       "      <td>[EC:1.2.1.11/4.3332]</td>\n",
       "      <td>EC:1.2.1.11</td>\n",
       "      <td>4.3332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q7UZH7</td>\n",
       "      <td>[EC:6.3.4.2/2.8527]</td>\n",
       "      <td>EC:6.3.4.2</td>\n",
       "      <td>2.8527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q7UZR6</td>\n",
       "      <td>[EC:6.3.2.1/5.8351]</td>\n",
       "      <td>EC:6.3.2.1</td>\n",
       "      <td>5.8351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q7V010</td>\n",
       "      <td>[EC:2.1.1.192/3.8433]</td>\n",
       "      <td>EC:2.1.1.192</td>\n",
       "      <td>3.8433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q7V0C4</td>\n",
       "      <td>[EC:2.7.11.1/7.2259]</td>\n",
       "      <td>EC:2.7.11.1</td>\n",
       "      <td>7.2259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1934</th>\n",
       "      <td>Q7V3R0</td>\n",
       "      <td>[EC:4.2.3.187/8.4480]</td>\n",
       "      <td>EC:4.2.3.187</td>\n",
       "      <td>8.4480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1935</th>\n",
       "      <td>Q7V3R1</td>\n",
       "      <td>[EC:2.7.1.66/8.4444]</td>\n",
       "      <td>EC:2.7.1.66</td>\n",
       "      <td>8.4444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1936</th>\n",
       "      <td>Q7V3R2</td>\n",
       "      <td>[EC:1.17.99.6/3.9525, EC:4.1.99.3/9.3495, EC:4...</td>\n",
       "      <td>EC:1.17.99.6, EC:4.1.99.3, EC:4.1.99.13</td>\n",
       "      <td>3.9525, 9.3495, 9.8844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937</th>\n",
       "      <td>Q7V3R3</td>\n",
       "      <td>[EC:3.5.2.6/7.2768, EC:3.1.1.107/8.2078]</td>\n",
       "      <td>EC:3.5.2.6, EC:3.1.1.107</td>\n",
       "      <td>7.2768, 8.2078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1938</th>\n",
       "      <td>Q7V3R6</td>\n",
       "      <td>[EC:6.2.1.57/9.1441, EC:2.4.1.33/9.5583]</td>\n",
       "      <td>EC:6.2.1.57, EC:2.4.1.33</td>\n",
       "      <td>9.1441, 9.5583</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1939 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     UniProt                                     CC pvalue full  \\\n",
       "0     Q7TU21                               [EC:1.2.1.11/4.3332]   \n",
       "1     Q7UZH7                                [EC:6.3.4.2/2.8527]   \n",
       "2     Q7UZR6                                [EC:6.3.2.1/5.8351]   \n",
       "3     Q7V010                              [EC:2.1.1.192/3.8433]   \n",
       "4     Q7V0C4                               [EC:2.7.11.1/7.2259]   \n",
       "...      ...                                                ...   \n",
       "1934  Q7V3R0                              [EC:4.2.3.187/8.4480]   \n",
       "1935  Q7V3R1                               [EC:2.7.1.66/8.4444]   \n",
       "1936  Q7V3R2  [EC:1.17.99.6/3.9525, EC:4.1.99.3/9.3495, EC:4...   \n",
       "1937  Q7V3R3           [EC:3.5.2.6/7.2768, EC:3.1.1.107/8.2078]   \n",
       "1938  Q7V3R6           [EC:6.2.1.57/9.1441, EC:2.4.1.33/9.5583]   \n",
       "\n",
       "                                 CC pvalue EC  CC pvalue cluster dist  \n",
       "0                                 EC:1.2.1.11                  4.3332  \n",
       "1                                  EC:6.3.4.2                  2.8527  \n",
       "2                                  EC:6.3.2.1                  5.8351  \n",
       "3                                EC:2.1.1.192                  3.8433  \n",
       "4                                 EC:2.7.11.1                  7.2259  \n",
       "...                                       ...                     ...  \n",
       "1934                             EC:4.2.3.187                  8.4480  \n",
       "1935                              EC:2.7.1.66                  8.4444  \n",
       "1936  EC:1.17.99.6, EC:4.1.99.3, EC:4.1.99.13  3.9525, 9.3495, 9.8844  \n",
       "1937                 EC:3.5.2.6, EC:3.1.1.107          7.2768, 8.2078  \n",
       "1938                 EC:6.2.1.57, EC:2.4.1.33          9.1441, 9.5583  \n",
       "\n",
       "[1939 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CLEAN-Contact pvalue results, with distance from cluster center\n",
    "cc_pvalue = process_results(cc_pvalue_path, ['UniProt', 'CC pvalue full'])\n",
    "cc_pvalue['CC pvalue EC'] = cc_pvalue['CC pvalue full'].apply(extract_ec_values)\n",
    "cc_pvalue['CC pvalue cluster dist'] = cc_pvalue['CC pvalue full'].apply(extract_from_list)\n",
    "cc_pvalue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "23307770",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The columns match. Proceeding to combine the DataFrames.\n"
     ]
    }
   ],
   "source": [
    "# combine and format the pvalue results if they have the same UniProt id and EC predictions\n",
    "if cc_pvalue_conf['CC pvalue conf EC'].equals(cc_pvalue['CC pvalue EC']):\n",
    "    print(\"The columns match. Proceeding to combine the DataFrames.\")\n",
    "    \n",
    "    # Combine the DataFrames on 'UniProt' and pvalue columns\n",
    "    cc_pvalue_results_full = pd.merge(cc_pvalue_conf, cc_pvalue, left_on=['UniProt', 'CC pvalue conf EC'], right_on=['UniProt', 'CC pvalue EC'])\n",
    "    \n",
    "    # Drop the redundant pvalue ec column \n",
    "    cc_pvalue_results_full.drop(columns=['CC pvalue conf EC'], inplace=True)\n",
    "    \n",
    "else:\n",
    "    print(\"The columns do not match.\")\n",
    "\n",
    "cc_pvalue_results_full.drop(columns=['CC pvalue conf full', 'CC pvalue full'], inplace=True)\n",
    "col_to_move = cc_pvalue_results_full.pop('CC pvalue EC')\n",
    "cc_pvalue_results_full.insert(3, 'CC pvalue EC', col_to_move)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b9b32ad4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>CC pvalue EC</th>\n",
       "      <th>CC pvalue conf</th>\n",
       "      <th>CC pvalue conf SD</th>\n",
       "      <th>CC pvalue cluster dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q7TU21</td>\n",
       "      <td>Aspartate-semialdehyde dehydrogenase OS=Prochl...</td>\n",
       "      <td>MRNSPFLPNRPLKVAVLGSSGAVGSELLKILEERDFPISELVLLSS...</td>\n",
       "      <td>EC:1.2.1.11</td>\n",
       "      <td>0.9863</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>4.3332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q7UZH7</td>\n",
       "      <td>CTP synthase OS=Prochlorococcus marinus subsp....</td>\n",
       "      <td>MSKFVFVTGGVVSSIGKGIVAASLGRLLKSRGYSVSILKLDPYLNV...</td>\n",
       "      <td>EC:6.3.4.2</td>\n",
       "      <td>0.9961</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>2.8527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q7UZR6</td>\n",
       "      <td>Bifunctional pantoate ligase/cytidylate kinase...</td>\n",
       "      <td>MNKIIIRKTEDLKEWRRNLKCDINFIPTMGNLHDGHQKLISTAQSS...</td>\n",
       "      <td>EC:6.3.2.1</td>\n",
       "      <td>0.8120</td>\n",
       "      <td>0.0551</td>\n",
       "      <td>5.8351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q7V010</td>\n",
       "      <td>Probable dual-specificity RNA methyltransferas...</td>\n",
       "      <td>MKNLLGCSVKDLEKIALNYGQAAFRGRQIYNWLYNYKNRSKSIDEI...</td>\n",
       "      <td>EC:2.1.1.192</td>\n",
       "      <td>0.9923</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>3.8433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q7V0C4</td>\n",
       "      <td>Circadian clock oscillator protein KaiC OS=Pro...</td>\n",
       "      <td>MKDKKISKSIKMQVQKIPTGIEGFDDVCRGGLPAARSTLVSGTSGT...</td>\n",
       "      <td>EC:2.7.11.1</td>\n",
       "      <td>0.0911</td>\n",
       "      <td>0.0452</td>\n",
       "      <td>7.2259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1934</th>\n",
       "      <td>Q7V3R0</td>\n",
       "      <td>Antitermination protein NusB OS=Prochlorococcu...</td>\n",
       "      <td>MMHNRSLSRELSLLSLGLIKDTADLELNKIQIDEIFESALDSLINH...</td>\n",
       "      <td>EC:4.2.3.187</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0012</td>\n",
       "      <td>8.4480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1935</th>\n",
       "      <td>Q7V3R1</td>\n",
       "      <td>Transporter OS=Prochlorococcus marinus subsp. ...</td>\n",
       "      <td>MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...</td>\n",
       "      <td>EC:2.7.1.66</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0012</td>\n",
       "      <td>8.4444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1936</th>\n",
       "      <td>Q7V3R2</td>\n",
       "      <td>4Fe-4S ferredoxin-type domain-containing prote...</td>\n",
       "      <td>MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...</td>\n",
       "      <td>EC:1.17.99.6, EC:4.1.99.3, EC:4.1.99.13</td>\n",
       "      <td>0.9914, 0.0000, 0.0000</td>\n",
       "      <td>0.0028, 0.0000, 0.0000</td>\n",
       "      <td>3.9525, 9.3495, 9.8844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937</th>\n",
       "      <td>Q7V3R3</td>\n",
       "      <td>Uncharacterized protein OS=Prochlorococcus mar...</td>\n",
       "      <td>MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...</td>\n",
       "      <td>EC:3.5.2.6, EC:3.1.1.107</td>\n",
       "      <td>0.0787, 0.0034</td>\n",
       "      <td>0.0403, 0.0027</td>\n",
       "      <td>7.2768, 8.2078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1938</th>\n",
       "      <td>Q7V3R6</td>\n",
       "      <td>PRC-barrel domain-containing protein OS=Prochl...</td>\n",
       "      <td>MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...</td>\n",
       "      <td>EC:6.2.1.57, EC:2.4.1.33</td>\n",
       "      <td>0.0001, 0.0000</td>\n",
       "      <td>0.0001, 0.0000</td>\n",
       "      <td>9.1441, 9.5583</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1939 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     UniProt                                         Annotation  \\\n",
       "0     Q7TU21  Aspartate-semialdehyde dehydrogenase OS=Prochl...   \n",
       "1     Q7UZH7  CTP synthase OS=Prochlorococcus marinus subsp....   \n",
       "2     Q7UZR6  Bifunctional pantoate ligase/cytidylate kinase...   \n",
       "3     Q7V010  Probable dual-specificity RNA methyltransferas...   \n",
       "4     Q7V0C4  Circadian clock oscillator protein KaiC OS=Pro...   \n",
       "...      ...                                                ...   \n",
       "1934  Q7V3R0  Antitermination protein NusB OS=Prochlorococcu...   \n",
       "1935  Q7V3R1  Transporter OS=Prochlorococcus marinus subsp. ...   \n",
       "1936  Q7V3R2  4Fe-4S ferredoxin-type domain-containing prote...   \n",
       "1937  Q7V3R3  Uncharacterized protein OS=Prochlorococcus mar...   \n",
       "1938  Q7V3R6  PRC-barrel domain-containing protein OS=Prochl...   \n",
       "\n",
       "                                               Sequence  \\\n",
       "0     MRNSPFLPNRPLKVAVLGSSGAVGSELLKILEERDFPISELVLLSS...   \n",
       "1     MSKFVFVTGGVVSSIGKGIVAASLGRLLKSRGYSVSILKLDPYLNV...   \n",
       "2     MNKIIIRKTEDLKEWRRNLKCDINFIPTMGNLHDGHQKLISTAQSS...   \n",
       "3     MKNLLGCSVKDLEKIALNYGQAAFRGRQIYNWLYNYKNRSKSIDEI...   \n",
       "4     MKDKKISKSIKMQVQKIPTGIEGFDDVCRGGLPAARSTLVSGTSGT...   \n",
       "...                                                 ...   \n",
       "1934  MMHNRSLSRELSLLSLGLIKDTADLELNKIQIDEIFESALDSLINH...   \n",
       "1935  MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...   \n",
       "1936  MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...   \n",
       "1937  MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...   \n",
       "1938  MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...   \n",
       "\n",
       "                                 CC pvalue EC          CC pvalue conf  \\\n",
       "0                                 EC:1.2.1.11                  0.9863   \n",
       "1                                  EC:6.3.4.2                  0.9961   \n",
       "2                                  EC:6.3.2.1                  0.8120   \n",
       "3                                EC:2.1.1.192                  0.9923   \n",
       "4                                 EC:2.7.11.1                  0.0911   \n",
       "...                                       ...                     ...   \n",
       "1934                             EC:4.2.3.187                  0.0014   \n",
       "1935                              EC:2.7.1.66                  0.0014   \n",
       "1936  EC:1.17.99.6, EC:4.1.99.3, EC:4.1.99.13  0.9914, 0.0000, 0.0000   \n",
       "1937                 EC:3.5.2.6, EC:3.1.1.107          0.0787, 0.0034   \n",
       "1938                 EC:6.2.1.57, EC:2.4.1.33          0.0001, 0.0000   \n",
       "\n",
       "           CC pvalue conf SD  CC pvalue cluster dist  \n",
       "0                     0.0041                  4.3332  \n",
       "1                     0.0017                  2.8527  \n",
       "2                     0.0551                  5.8351  \n",
       "3                     0.0025                  3.8433  \n",
       "4                     0.0452                  7.2259  \n",
       "...                      ...                     ...  \n",
       "1934                  0.0012                  8.4480  \n",
       "1935                  0.0012                  8.4444  \n",
       "1936  0.0028, 0.0000, 0.0000  3.9525, 9.3495, 9.8844  \n",
       "1937          0.0403, 0.0027          7.2768, 8.2078  \n",
       "1938          0.0001, 0.0000          9.1441, 9.5583  \n",
       "\n",
       "[1939 rows x 7 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#cc_pvalue_results_full.to_csv('cc_pvalue_results_full.csv', index=False)\n",
    "cc_pvalue_results_full"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60fbcc56",
   "metadata": {},
   "source": [
    "## CLEAN\n",
    "CLEAN only provides the GMM option for the maxsep inference method, not pvalue, and those results are not in addition to the maxsep method results without the GMM. For the case study each inference method was run once, with the GMM if possible. This is why the CLEAN results start with 1 dataset for maxsep (only with GMM) and 1 dataset for pvalue (no GMM option available). Also note that CLEAN's GMM results only include the confidence score; no standard deviation for the confidence score.\n",
    "\n",
    "Maxsep with GMM\n",
    "* results from using the maxsep inference method, with the GMM to get confidence scores\n",
    "\n",
    "Pvalue\n",
    "* results from using the pvalue inference method, without the GMM since there's not a GMM option available \n",
    "\n",
    "Combined\n",
    "* combine the maxsep and pvalue results and share in this directory as **clean_combined_results_full.csv**\n",
    "    * 1942 rows of data\n",
    "    * 7 columns\n",
    "        * 'UniProt': UniProt id\n",
    "        * 'Annotation': header annotations from the fasta file\n",
    "        * 'Sequence': protein sequences from the fasta file\n",
    "        * 'Clean maxsep EC': CLEAN's EC predictions from using the maxsep inference method\n",
    "        * 'Clean maxsep conf': CLEAN's confidence scores from using the maxsep inference method with the GMM\n",
    "        * 'Clean pvalue EC': CLEAN's EC predictions from using the pvalue inference method\n",
    "        * 'Clean pvalue cluster dist': CLEAN's EC prediction's distance from the cluster center from using the pvalue inference method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bbbc531e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>Clean maxsep conf full</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Clean maxsep EC</th>\n",
       "      <th>Clean maxsep conf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q7TU21</td>\n",
       "      <td>[EC:1.2.1.11/0.9935]</td>\n",
       "      <td>Aspartate-semialdehyde dehydrogenase OS=Prochl...</td>\n",
       "      <td>MRNSPFLPNRPLKVAVLGSSGAVGSELLKILEERDFPISELVLLSS...</td>\n",
       "      <td>EC:1.2.1.11</td>\n",
       "      <td>0.9935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q7UZH7</td>\n",
       "      <td>[EC:6.3.4.2/0.9976]</td>\n",
       "      <td>CTP synthase OS=Prochlorococcus marinus subsp....</td>\n",
       "      <td>MSKFVFVTGGVVSSIGKGIVAASLGRLLKSRGYSVSILKLDPYLNV...</td>\n",
       "      <td>EC:6.3.4.2</td>\n",
       "      <td>0.9976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q7UZR6</td>\n",
       "      <td>[EC:2.7.4.25/0.2534]</td>\n",
       "      <td>Bifunctional pantoate ligase/cytidylate kinase...</td>\n",
       "      <td>MNKIIIRKTEDLKEWRRNLKCDINFIPTMGNLHDGHQKLISTAQSS...</td>\n",
       "      <td>EC:2.7.4.25</td>\n",
       "      <td>0.2534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q7V010</td>\n",
       "      <td>[EC:2.1.1.192/0.9917]</td>\n",
       "      <td>Probable dual-specificity RNA methyltransferas...</td>\n",
       "      <td>MKNLLGCSVKDLEKIALNYGQAAFRGRQIYNWLYNYKNRSKSIDEI...</td>\n",
       "      <td>EC:2.1.1.192</td>\n",
       "      <td>0.9917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q7V0C4</td>\n",
       "      <td>[EC:2.7.11.1/0.0451]</td>\n",
       "      <td>Circadian clock oscillator protein KaiC OS=Pro...</td>\n",
       "      <td>MKDKKISKSIKMQVQKIPTGIEGFDDVCRGGLPAARSTLVSGTSGT...</td>\n",
       "      <td>EC:2.7.11.1</td>\n",
       "      <td>0.0451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937</th>\n",
       "      <td>Q7V3R0</td>\n",
       "      <td>[EC:2.5.1.17/0.0002]</td>\n",
       "      <td>Antitermination protein NusB OS=Prochlorococcu...</td>\n",
       "      <td>MMHNRSLSRELSLLSLGLIKDTADLELNKIQIDEIFESALDSLINH...</td>\n",
       "      <td>EC:2.5.1.17</td>\n",
       "      <td>0.0002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1938</th>\n",
       "      <td>Q7V3R1</td>\n",
       "      <td>[EC:1.1.5.2/0.0001]</td>\n",
       "      <td>Transporter OS=Prochlorococcus marinus subsp. ...</td>\n",
       "      <td>MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...</td>\n",
       "      <td>EC:1.1.5.2</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939</th>\n",
       "      <td>Q7V3R2</td>\n",
       "      <td>[EC:1.17.99.6/0.9886]</td>\n",
       "      <td>4Fe-4S ferredoxin-type domain-containing prote...</td>\n",
       "      <td>MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...</td>\n",
       "      <td>EC:1.17.99.6</td>\n",
       "      <td>0.9886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>Q7V3R3</td>\n",
       "      <td>[EC:5.2.1.8/0.0077]</td>\n",
       "      <td>Uncharacterized protein OS=Prochlorococcus mar...</td>\n",
       "      <td>MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...</td>\n",
       "      <td>EC:5.2.1.8</td>\n",
       "      <td>0.0077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941</th>\n",
       "      <td>Q7V3R6</td>\n",
       "      <td>[EC:3.2.2.n1/0.0007]</td>\n",
       "      <td>PRC-barrel domain-containing protein OS=Prochl...</td>\n",
       "      <td>MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...</td>\n",
       "      <td>EC:3.2.2.n1</td>\n",
       "      <td>0.0007</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1942 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     UniProt Clean maxsep conf full  \\\n",
       "0     Q7TU21   [EC:1.2.1.11/0.9935]   \n",
       "1     Q7UZH7    [EC:6.3.4.2/0.9976]   \n",
       "2     Q7UZR6   [EC:2.7.4.25/0.2534]   \n",
       "3     Q7V010  [EC:2.1.1.192/0.9917]   \n",
       "4     Q7V0C4   [EC:2.7.11.1/0.0451]   \n",
       "...      ...                    ...   \n",
       "1937  Q7V3R0   [EC:2.5.1.17/0.0002]   \n",
       "1938  Q7V3R1    [EC:1.1.5.2/0.0001]   \n",
       "1939  Q7V3R2  [EC:1.17.99.6/0.9886]   \n",
       "1940  Q7V3R3    [EC:5.2.1.8/0.0077]   \n",
       "1941  Q7V3R6   [EC:3.2.2.n1/0.0007]   \n",
       "\n",
       "                                             Annotation  \\\n",
       "0     Aspartate-semialdehyde dehydrogenase OS=Prochl...   \n",
       "1     CTP synthase OS=Prochlorococcus marinus subsp....   \n",
       "2     Bifunctional pantoate ligase/cytidylate kinase...   \n",
       "3     Probable dual-specificity RNA methyltransferas...   \n",
       "4     Circadian clock oscillator protein KaiC OS=Pro...   \n",
       "...                                                 ...   \n",
       "1937  Antitermination protein NusB OS=Prochlorococcu...   \n",
       "1938  Transporter OS=Prochlorococcus marinus subsp. ...   \n",
       "1939  4Fe-4S ferredoxin-type domain-containing prote...   \n",
       "1940  Uncharacterized protein OS=Prochlorococcus mar...   \n",
       "1941  PRC-barrel domain-containing protein OS=Prochl...   \n",
       "\n",
       "                                               Sequence Clean maxsep EC  \\\n",
       "0     MRNSPFLPNRPLKVAVLGSSGAVGSELLKILEERDFPISELVLLSS...     EC:1.2.1.11   \n",
       "1     MSKFVFVTGGVVSSIGKGIVAASLGRLLKSRGYSVSILKLDPYLNV...      EC:6.3.4.2   \n",
       "2     MNKIIIRKTEDLKEWRRNLKCDINFIPTMGNLHDGHQKLISTAQSS...     EC:2.7.4.25   \n",
       "3     MKNLLGCSVKDLEKIALNYGQAAFRGRQIYNWLYNYKNRSKSIDEI...    EC:2.1.1.192   \n",
       "4     MKDKKISKSIKMQVQKIPTGIEGFDDVCRGGLPAARSTLVSGTSGT...     EC:2.7.11.1   \n",
       "...                                                 ...             ...   \n",
       "1937  MMHNRSLSRELSLLSLGLIKDTADLELNKIQIDEIFESALDSLINH...     EC:2.5.1.17   \n",
       "1938  MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...      EC:1.1.5.2   \n",
       "1939  MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...    EC:1.17.99.6   \n",
       "1940  MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...      EC:5.2.1.8   \n",
       "1941  MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...     EC:3.2.2.n1   \n",
       "\n",
       "     Clean maxsep conf  \n",
       "0               0.9935  \n",
       "1               0.9976  \n",
       "2               0.2534  \n",
       "3               0.9917  \n",
       "4               0.0451  \n",
       "...                ...  \n",
       "1937            0.0002  \n",
       "1938            0.0001  \n",
       "1939            0.9886  \n",
       "1940            0.0077  \n",
       "1941            0.0007  \n",
       "\n",
       "[1942 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CLEAN's maxsep results, with confidence scores from using the gmm\n",
    "clean_maxsep_conf = process_results(clean_maxsep_conf_path, ['UniProt', 'Clean maxsep conf full'])\n",
    "clean_maxsep_conf['UniProt'] = clean_maxsep_conf['UniProt'].apply(remove_extra_id_format)\n",
    "clean_maxsep_conf = update_dataframe_with_annotations(med4_fasta_path, clean_maxsep_conf)\n",
    "clean_maxsep_conf['Clean maxsep EC'] = clean_maxsep_conf['Clean maxsep conf full'].apply(extract_ec_values)\n",
    "clean_maxsep_conf['Clean maxsep conf'] = clean_maxsep_conf['Clean maxsep conf full'].apply(extract_from_list)\n",
    "clean_maxsep_conf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "63abe1cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>Clean pvalue full</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Clean pvalue EC</th>\n",
       "      <th>Clean pvalue cluster dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q7TU21</td>\n",
       "      <td>[EC:1.2.1.11/3.7585]</td>\n",
       "      <td>Aspartate-semialdehyde dehydrogenase OS=Prochl...</td>\n",
       "      <td>MRNSPFLPNRPLKVAVLGSSGAVGSELLKILEERDFPISELVLLSS...</td>\n",
       "      <td>EC:1.2.1.11</td>\n",
       "      <td>3.7585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q7UZH7</td>\n",
       "      <td>[EC:6.3.4.2/2.9574]</td>\n",
       "      <td>CTP synthase OS=Prochlorococcus marinus subsp....</td>\n",
       "      <td>MSKFVFVTGGVVSSIGKGIVAASLGRLLKSRGYSVSILKLDPYLNV...</td>\n",
       "      <td>EC:6.3.4.2</td>\n",
       "      <td>2.9574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q7UZR6</td>\n",
       "      <td>[EC:2.7.4.25/6.4352]</td>\n",
       "      <td>Bifunctional pantoate ligase/cytidylate kinase...</td>\n",
       "      <td>MNKIIIRKTEDLKEWRRNLKCDINFIPTMGNLHDGHQKLISTAQSS...</td>\n",
       "      <td>EC:2.7.4.25</td>\n",
       "      <td>6.4352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q7V010</td>\n",
       "      <td>[EC:2.1.1.192/3.9242, EC:2.1.1.224/4.9630]</td>\n",
       "      <td>Probable dual-specificity RNA methyltransferas...</td>\n",
       "      <td>MKNLLGCSVKDLEKIALNYGQAAFRGRQIYNWLYNYKNRSKSIDEI...</td>\n",
       "      <td>EC:2.1.1.192, EC:2.1.1.224</td>\n",
       "      <td>3.9242, 4.9630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q7V0C4</td>\n",
       "      <td>[EC:2.7.11.1/7.0824]</td>\n",
       "      <td>Circadian clock oscillator protein KaiC OS=Pro...</td>\n",
       "      <td>MKDKKISKSIKMQVQKIPTGIEGFDDVCRGGLPAARSTLVSGTSGT...</td>\n",
       "      <td>EC:2.7.11.1</td>\n",
       "      <td>7.0824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937</th>\n",
       "      <td>Q7V3R0</td>\n",
       "      <td>[EC:2.5.1.17/8.6199]</td>\n",
       "      <td>Antitermination protein NusB OS=Prochlorococcu...</td>\n",
       "      <td>MMHNRSLSRELSLLSLGLIKDTADLELNKIQIDEIFESALDSLINH...</td>\n",
       "      <td>EC:2.5.1.17</td>\n",
       "      <td>8.6199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1938</th>\n",
       "      <td>Q7V3R1</td>\n",
       "      <td>[EC:1.1.5.2/8.6942]</td>\n",
       "      <td>Transporter OS=Prochlorococcus marinus subsp. ...</td>\n",
       "      <td>MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...</td>\n",
       "      <td>EC:1.1.5.2</td>\n",
       "      <td>8.6942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939</th>\n",
       "      <td>Q7V3R2</td>\n",
       "      <td>[EC:1.17.99.6/4.1148]</td>\n",
       "      <td>4Fe-4S ferredoxin-type domain-containing prote...</td>\n",
       "      <td>MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...</td>\n",
       "      <td>EC:1.17.99.6</td>\n",
       "      <td>4.1148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>Q7V3R3</td>\n",
       "      <td>[EC:5.2.1.8/7.6222]</td>\n",
       "      <td>Uncharacterized protein OS=Prochlorococcus mar...</td>\n",
       "      <td>MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...</td>\n",
       "      <td>EC:5.2.1.8</td>\n",
       "      <td>7.6222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941</th>\n",
       "      <td>Q7V3R6</td>\n",
       "      <td>[EC:3.2.2.n1/8.2983]</td>\n",
       "      <td>PRC-barrel domain-containing protein OS=Prochl...</td>\n",
       "      <td>MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...</td>\n",
       "      <td>EC:3.2.2.n1</td>\n",
       "      <td>8.2983</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1942 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     UniProt                           Clean pvalue full  \\\n",
       "0     Q7TU21                        [EC:1.2.1.11/3.7585]   \n",
       "1     Q7UZH7                         [EC:6.3.4.2/2.9574]   \n",
       "2     Q7UZR6                        [EC:2.7.4.25/6.4352]   \n",
       "3     Q7V010  [EC:2.1.1.192/3.9242, EC:2.1.1.224/4.9630]   \n",
       "4     Q7V0C4                        [EC:2.7.11.1/7.0824]   \n",
       "...      ...                                         ...   \n",
       "1937  Q7V3R0                        [EC:2.5.1.17/8.6199]   \n",
       "1938  Q7V3R1                         [EC:1.1.5.2/8.6942]   \n",
       "1939  Q7V3R2                       [EC:1.17.99.6/4.1148]   \n",
       "1940  Q7V3R3                         [EC:5.2.1.8/7.6222]   \n",
       "1941  Q7V3R6                        [EC:3.2.2.n1/8.2983]   \n",
       "\n",
       "                                             Annotation  \\\n",
       "0     Aspartate-semialdehyde dehydrogenase OS=Prochl...   \n",
       "1     CTP synthase OS=Prochlorococcus marinus subsp....   \n",
       "2     Bifunctional pantoate ligase/cytidylate kinase...   \n",
       "3     Probable dual-specificity RNA methyltransferas...   \n",
       "4     Circadian clock oscillator protein KaiC OS=Pro...   \n",
       "...                                                 ...   \n",
       "1937  Antitermination protein NusB OS=Prochlorococcu...   \n",
       "1938  Transporter OS=Prochlorococcus marinus subsp. ...   \n",
       "1939  4Fe-4S ferredoxin-type domain-containing prote...   \n",
       "1940  Uncharacterized protein OS=Prochlorococcus mar...   \n",
       "1941  PRC-barrel domain-containing protein OS=Prochl...   \n",
       "\n",
       "                                               Sequence  \\\n",
       "0     MRNSPFLPNRPLKVAVLGSSGAVGSELLKILEERDFPISELVLLSS...   \n",
       "1     MSKFVFVTGGVVSSIGKGIVAASLGRLLKSRGYSVSILKLDPYLNV...   \n",
       "2     MNKIIIRKTEDLKEWRRNLKCDINFIPTMGNLHDGHQKLISTAQSS...   \n",
       "3     MKNLLGCSVKDLEKIALNYGQAAFRGRQIYNWLYNYKNRSKSIDEI...   \n",
       "4     MKDKKISKSIKMQVQKIPTGIEGFDDVCRGGLPAARSTLVSGTSGT...   \n",
       "...                                                 ...   \n",
       "1937  MMHNRSLSRELSLLSLGLIKDTADLELNKIQIDEIFESALDSLINH...   \n",
       "1938  MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...   \n",
       "1939  MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...   \n",
       "1940  MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...   \n",
       "1941  MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...   \n",
       "\n",
       "                 Clean pvalue EC Clean pvalue cluster dist  \n",
       "0                    EC:1.2.1.11                    3.7585  \n",
       "1                     EC:6.3.4.2                    2.9574  \n",
       "2                    EC:2.7.4.25                    6.4352  \n",
       "3     EC:2.1.1.192, EC:2.1.1.224            3.9242, 4.9630  \n",
       "4                    EC:2.7.11.1                    7.0824  \n",
       "...                          ...                       ...  \n",
       "1937                 EC:2.5.1.17                    8.6199  \n",
       "1938                  EC:1.1.5.2                    8.6942  \n",
       "1939                EC:1.17.99.6                    4.1148  \n",
       "1940                  EC:5.2.1.8                    7.6222  \n",
       "1941                 EC:3.2.2.n1                    8.2983  \n",
       "\n",
       "[1942 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CLEAN's pvalue results, with distance from cluster center\n",
    "clean_pvalue = process_results(clean_pvalue_path, ['UniProt', 'Clean pvalue full'])\n",
    "clean_pvalue['UniProt'] = clean_pvalue['UniProt'].apply(remove_extra_id_format)\n",
    "clean_pvalue = update_dataframe_with_annotations(med4_fasta_path, clean_pvalue)\n",
    "clean_pvalue['Clean pvalue EC'] = clean_pvalue['Clean pvalue full'].apply(extract_ec_values)\n",
    "clean_pvalue['Clean pvalue cluster dist'] = clean_pvalue['Clean pvalue full'].apply(extract_from_list)\n",
    "clean_pvalue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "947aa9e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Clean maxsep EC</th>\n",
       "      <th>Clean maxsep conf</th>\n",
       "      <th>Clean pvalue EC</th>\n",
       "      <th>Clean pvalue cluster dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q7TU21</td>\n",
       "      <td>Aspartate-semialdehyde dehydrogenase OS=Prochl...</td>\n",
       "      <td>MRNSPFLPNRPLKVAVLGSSGAVGSELLKILEERDFPISELVLLSS...</td>\n",
       "      <td>EC:1.2.1.11</td>\n",
       "      <td>0.9935</td>\n",
       "      <td>EC:1.2.1.11</td>\n",
       "      <td>3.7585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q7UZH7</td>\n",
       "      <td>CTP synthase OS=Prochlorococcus marinus subsp....</td>\n",
       "      <td>MSKFVFVTGGVVSSIGKGIVAASLGRLLKSRGYSVSILKLDPYLNV...</td>\n",
       "      <td>EC:6.3.4.2</td>\n",
       "      <td>0.9976</td>\n",
       "      <td>EC:6.3.4.2</td>\n",
       "      <td>2.9574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q7UZR6</td>\n",
       "      <td>Bifunctional pantoate ligase/cytidylate kinase...</td>\n",
       "      <td>MNKIIIRKTEDLKEWRRNLKCDINFIPTMGNLHDGHQKLISTAQSS...</td>\n",
       "      <td>EC:2.7.4.25</td>\n",
       "      <td>0.2534</td>\n",
       "      <td>EC:2.7.4.25</td>\n",
       "      <td>6.4352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q7V010</td>\n",
       "      <td>Probable dual-specificity RNA methyltransferas...</td>\n",
       "      <td>MKNLLGCSVKDLEKIALNYGQAAFRGRQIYNWLYNYKNRSKSIDEI...</td>\n",
       "      <td>EC:2.1.1.192</td>\n",
       "      <td>0.9917</td>\n",
       "      <td>EC:2.1.1.192, EC:2.1.1.224</td>\n",
       "      <td>3.9242, 4.9630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q7V0C4</td>\n",
       "      <td>Circadian clock oscillator protein KaiC OS=Pro...</td>\n",
       "      <td>MKDKKISKSIKMQVQKIPTGIEGFDDVCRGGLPAARSTLVSGTSGT...</td>\n",
       "      <td>EC:2.7.11.1</td>\n",
       "      <td>0.0451</td>\n",
       "      <td>EC:2.7.11.1</td>\n",
       "      <td>7.0824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937</th>\n",
       "      <td>Q7V3R0</td>\n",
       "      <td>Antitermination protein NusB OS=Prochlorococcu...</td>\n",
       "      <td>MMHNRSLSRELSLLSLGLIKDTADLELNKIQIDEIFESALDSLINH...</td>\n",
       "      <td>EC:2.5.1.17</td>\n",
       "      <td>0.0002</td>\n",
       "      <td>EC:2.5.1.17</td>\n",
       "      <td>8.6199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1938</th>\n",
       "      <td>Q7V3R1</td>\n",
       "      <td>Transporter OS=Prochlorococcus marinus subsp. ...</td>\n",
       "      <td>MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...</td>\n",
       "      <td>EC:1.1.5.2</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>EC:1.1.5.2</td>\n",
       "      <td>8.6942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939</th>\n",
       "      <td>Q7V3R2</td>\n",
       "      <td>4Fe-4S ferredoxin-type domain-containing prote...</td>\n",
       "      <td>MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...</td>\n",
       "      <td>EC:1.17.99.6</td>\n",
       "      <td>0.9886</td>\n",
       "      <td>EC:1.17.99.6</td>\n",
       "      <td>4.1148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>Q7V3R3</td>\n",
       "      <td>Uncharacterized protein OS=Prochlorococcus mar...</td>\n",
       "      <td>MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...</td>\n",
       "      <td>EC:5.2.1.8</td>\n",
       "      <td>0.0077</td>\n",
       "      <td>EC:5.2.1.8</td>\n",
       "      <td>7.6222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941</th>\n",
       "      <td>Q7V3R6</td>\n",
       "      <td>PRC-barrel domain-containing protein OS=Prochl...</td>\n",
       "      <td>MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...</td>\n",
       "      <td>EC:3.2.2.n1</td>\n",
       "      <td>0.0007</td>\n",
       "      <td>EC:3.2.2.n1</td>\n",
       "      <td>8.2983</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1942 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     UniProt                                         Annotation  \\\n",
       "0     Q7TU21  Aspartate-semialdehyde dehydrogenase OS=Prochl...   \n",
       "1     Q7UZH7  CTP synthase OS=Prochlorococcus marinus subsp....   \n",
       "2     Q7UZR6  Bifunctional pantoate ligase/cytidylate kinase...   \n",
       "3     Q7V010  Probable dual-specificity RNA methyltransferas...   \n",
       "4     Q7V0C4  Circadian clock oscillator protein KaiC OS=Pro...   \n",
       "...      ...                                                ...   \n",
       "1937  Q7V3R0  Antitermination protein NusB OS=Prochlorococcu...   \n",
       "1938  Q7V3R1  Transporter OS=Prochlorococcus marinus subsp. ...   \n",
       "1939  Q7V3R2  4Fe-4S ferredoxin-type domain-containing prote...   \n",
       "1940  Q7V3R3  Uncharacterized protein OS=Prochlorococcus mar...   \n",
       "1941  Q7V3R6  PRC-barrel domain-containing protein OS=Prochl...   \n",
       "\n",
       "                                               Sequence Clean maxsep EC  \\\n",
       "0     MRNSPFLPNRPLKVAVLGSSGAVGSELLKILEERDFPISELVLLSS...     EC:1.2.1.11   \n",
       "1     MSKFVFVTGGVVSSIGKGIVAASLGRLLKSRGYSVSILKLDPYLNV...      EC:6.3.4.2   \n",
       "2     MNKIIIRKTEDLKEWRRNLKCDINFIPTMGNLHDGHQKLISTAQSS...     EC:2.7.4.25   \n",
       "3     MKNLLGCSVKDLEKIALNYGQAAFRGRQIYNWLYNYKNRSKSIDEI...    EC:2.1.1.192   \n",
       "4     MKDKKISKSIKMQVQKIPTGIEGFDDVCRGGLPAARSTLVSGTSGT...     EC:2.7.11.1   \n",
       "...                                                 ...             ...   \n",
       "1937  MMHNRSLSRELSLLSLGLIKDTADLELNKIQIDEIFESALDSLINH...     EC:2.5.1.17   \n",
       "1938  MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...      EC:1.1.5.2   \n",
       "1939  MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...    EC:1.17.99.6   \n",
       "1940  MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...      EC:5.2.1.8   \n",
       "1941  MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...     EC:3.2.2.n1   \n",
       "\n",
       "     Clean maxsep conf             Clean pvalue EC Clean pvalue cluster dist  \n",
       "0               0.9935                 EC:1.2.1.11                    3.7585  \n",
       "1               0.9976                  EC:6.3.4.2                    2.9574  \n",
       "2               0.2534                 EC:2.7.4.25                    6.4352  \n",
       "3               0.9917  EC:2.1.1.192, EC:2.1.1.224            3.9242, 4.9630  \n",
       "4               0.0451                 EC:2.7.11.1                    7.0824  \n",
       "...                ...                         ...                       ...  \n",
       "1937            0.0002                 EC:2.5.1.17                    8.6199  \n",
       "1938            0.0001                  EC:1.1.5.2                    8.6942  \n",
       "1939            0.9886                EC:1.17.99.6                    4.1148  \n",
       "1940            0.0077                  EC:5.2.1.8                    7.6222  \n",
       "1941            0.0007                 EC:3.2.2.n1                    8.2983  \n",
       "\n",
       "[1942 rows x 7 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine clean results from maxsep and pvalue\n",
    "clean_combined_results_full = pd.merge(clean_maxsep_conf, clean_pvalue, on=['UniProt', 'Annotation', 'Sequence'])\n",
    "clean_combined_results_full.drop(columns=['Clean maxsep conf full', 'Clean pvalue full'], inplace=True)\n",
    "#clean_combined_results_full.to_csv('clean_combined_results_full.csv', index=False)\n",
    "clean_combined_results_full  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4870275",
   "metadata": {},
   "source": [
    "## BioCyc\n",
    "BioCyc data on Prochlorococcus marinus subsp. pastoris str. CCMP1986, also known as MED4.\n",
    "\n",
    "Found and fixed an error where the UniProt id 'Q7V3L5' was incorrectly duplicated. One entry was correct, but the other was supposed to be 'Q7V2M2'. This was updated here in our data as well as reported to BioCyc, where the data has been updated in their records. This processing step also updated the EC number column with the EC number formatting to match CLEAN and CLEAN-Contact's EC number formatting.\n",
    "\n",
    "Update and share in this directory as **biocyc_updated.csv**.\n",
    "   * 728 rows of data\n",
    "   * 7 columns\n",
    "       * 'UniProt': UniProt id\n",
    "       * 'Genes of a reaction from All instances of Reactions in Prochlorococcus marinus pastoris CCMP1986': genes involved in these reactions\n",
    "       * 'Matches': reaction matches\n",
    "       * 'Accession-1': accession number from another database\n",
    "       * 'Accession-2': accession number from another database\n",
    "       * 'Gene products (polypeptides and RNAs only)': the gene products encoded by the genes in these reactions\n",
    "       * 'BioCyc EC': BioCyc's EC number, formatted to match CLEAN and CLEAN-Contact\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b82b2785",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "UniProt\n",
       "Q7V3L5    2\n",
       "Q7V2E7    1\n",
       "Q7TU77    1\n",
       "Q7V189    1\n",
       "Q7V177    1\n",
       "         ..\n",
       "Q7V2Q1    1\n",
       "Q7V2G7    1\n",
       "Q7V278    1\n",
       "Q7V121    1\n",
       "Q7V116    1\n",
       "Name: count, Length: 720, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biocyc = pd.read_csv(biocyc_path, delimiter='\\t')\n",
    "biocyc['UniProt'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "07715328",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>Genes of a reaction from All instances of Reactions in Prochlorococcus marinus pastoris CCMP1986</th>\n",
       "      <th>Matches</th>\n",
       "      <th>EC-Number</th>\n",
       "      <th>Accession-1</th>\n",
       "      <th>Accession-2</th>\n",
       "      <th>Gene products (polypeptides and RNAs only)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q7V3L5</td>\n",
       "      <td>accC</td>\n",
       "      <td>a [biotin carboxyl-carrier-protein dimer]-N6-b...</td>\n",
       "      <td>EC-6.3.4.14</td>\n",
       "      <td>PMM0060</td>\n",
       "      <td>TX50_RS00320</td>\n",
       "      <td>acetyl-CoA carboxylase biotin carboxylase subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>Q7V2M2</td>\n",
       "      <td>fabG</td>\n",
       "      <td>3-oxo-meristoyl-[acyl-carrier protein] reducta...</td>\n",
       "      <td>EC-1.1.1.100</td>\n",
       "      <td>PMM0453</td>\n",
       "      <td>TX50_RS02440</td>\n",
       "      <td>3-oxoacyl-[acyl-carrier-protein] reductase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    UniProt  \\\n",
       "3    Q7V3L5   \n",
       "140  Q7V2M2   \n",
       "\n",
       "    Genes of a reaction from All instances of Reactions in Prochlorococcus marinus pastoris CCMP1986  \\\n",
       "3                                                 accC                                                 \n",
       "140                                               fabG                                                 \n",
       "\n",
       "                                               Matches     EC-Number  \\\n",
       "3    a [biotin carboxyl-carrier-protein dimer]-N6-b...   EC-6.3.4.14   \n",
       "140  3-oxo-meristoyl-[acyl-carrier protein] reducta...  EC-1.1.1.100   \n",
       "\n",
       "    Accession-1   Accession-2  \\\n",
       "3       PMM0060  TX50_RS00320   \n",
       "140     PMM0453  TX50_RS02440   \n",
       "\n",
       "            Gene products (polypeptides and RNAs only)  \n",
       "3    acetyl-CoA carboxylase biotin carboxylase subunit  \n",
       "140         3-oxoacyl-[acyl-carrier-protein] reductase  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# correct the duplicated UniProt id\n",
    "biocyc_updated = biocyc.copy()\n",
    "biocyc_updated.loc[(biocyc_updated['UniProt'] == 'Q7V3L5') & (biocyc_updated['Genes of a reaction from All instances of Reactions in Prochlorococcus marinus pastoris CCMP1986'] == 'fabG'), 'UniProt'] = 'Q7V2M2'\n",
    "\n",
    "# view these changes\n",
    "biocyc_updated[biocyc_updated['UniProt'].isin(['Q7V3L5','Q7V2M2'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "046e3c64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>Genes of a reaction from All instances of Reactions in Prochlorococcus marinus pastoris CCMP1986</th>\n",
       "      <th>Matches</th>\n",
       "      <th>Accession-1</th>\n",
       "      <th>Accession-2</th>\n",
       "      <th>Gene products (polypeptides and RNAs only)</th>\n",
       "      <th>BioCyc EC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q7V2E7</td>\n",
       "      <td>aarA</td>\n",
       "      <td>octadecanal + acyl-carrier protein + NAD(P)+  ...</td>\n",
       "      <td>PMM0533</td>\n",
       "      <td>TX50_RS02855</td>\n",
       "      <td>long-chain acyl-[acyl-carrier-protein] reductase</td>\n",
       "      <td>EC:1.2.1.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q7V2E6</td>\n",
       "      <td>accA</td>\n",
       "      <td>acetyl-CoA + a [carboxyl-carrier protein dimer...</td>\n",
       "      <td>PMM0534</td>\n",
       "      <td>TX50_RS02860</td>\n",
       "      <td>acetyl-CoA carboxylase carboxyltransferase sub...</td>\n",
       "      <td>EC:2.1.3.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q7TUH9</td>\n",
       "      <td>accB</td>\n",
       "      <td>ATP + acetyl-CoA + hydrogencarbonate  -&gt;  ADP ...</td>\n",
       "      <td>PMM0027</td>\n",
       "      <td>TX50_RS00150</td>\n",
       "      <td>acetyl-CoA carboxylase biotin carboxyl carrier...</td>\n",
       "      <td>EC:6.4.1.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q7V3L5</td>\n",
       "      <td>accC</td>\n",
       "      <td>a [biotin carboxyl-carrier-protein dimer]-N6-b...</td>\n",
       "      <td>PMM0060</td>\n",
       "      <td>TX50_RS00320</td>\n",
       "      <td>acetyl-CoA carboxylase biotin carboxylase subunit</td>\n",
       "      <td>EC:6.3.4.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q7V1S1</td>\n",
       "      <td>accD</td>\n",
       "      <td>acetyl-CoA + a [carboxyl-carrier protein dimer...</td>\n",
       "      <td>PMM0784</td>\n",
       "      <td>TX50_RS04185</td>\n",
       "      <td>acetyl-CoA carboxylase carboxyltransferase sub...</td>\n",
       "      <td>EC:2.1.3.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ycf12</td>\n",
       "      <td>RXN-15490 // oxygen[in] + 4 H+[in] + 4 e-[memb...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>photosystem II reaction center protein Ycf12</td>\n",
       "      <td>EC:1.1.3, EC:1.10.3.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>Q7V0W4</td>\n",
       "      <td>ychF</td>\n",
       "      <td>ATP + H2O  -&gt;  ADP + phosphate + H+</td>\n",
       "      <td>PMM1138</td>\n",
       "      <td>TX50_RS06135</td>\n",
       "      <td>redox-regulated ATPase YchF</td>\n",
       "      <td>EC:3.6.1.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>725</th>\n",
       "      <td>Q7V3G0</td>\n",
       "      <td>zds</td>\n",
       "      <td>beta-zeacarotene + an electron-transfer quinon...</td>\n",
       "      <td>PMM0115</td>\n",
       "      <td>TX50_RS00605</td>\n",
       "      <td>9,9'-di-cis-zeta-carotene desaturase</td>\n",
       "      <td>EC:1.3.5.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>726</th>\n",
       "      <td>Q7V3E5</td>\n",
       "      <td>zntA</td>\n",
       "      <td>a metal cation[out] + ATP + H2O  -&gt;  a metal c...</td>\n",
       "      <td>PMM0131</td>\n",
       "      <td>TX50_RS00690</td>\n",
       "      <td>P-type cation transporter</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>727</th>\n",
       "      <td>Q7V116</td>\n",
       "      <td>zwf</td>\n",
       "      <td>D-glucopyranose 6-phosphate + NADP+  -&gt;  6-pho...</td>\n",
       "      <td>PMM1074</td>\n",
       "      <td>TX50_RS05765</td>\n",
       "      <td>glucose-6-phosphate dehydrogenase</td>\n",
       "      <td>EC:1.1.1.49</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>728 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    UniProt  \\\n",
       "0    Q7V2E7   \n",
       "1    Q7V2E6   \n",
       "2    Q7TUH9   \n",
       "3    Q7V3L5   \n",
       "4    Q7V1S1   \n",
       "..      ...   \n",
       "723     NaN   \n",
       "724  Q7V0W4   \n",
       "725  Q7V3G0   \n",
       "726  Q7V3E5   \n",
       "727  Q7V116   \n",
       "\n",
       "    Genes of a reaction from All instances of Reactions in Prochlorococcus marinus pastoris CCMP1986  \\\n",
       "0                                                 aarA                                                 \n",
       "1                                                 accA                                                 \n",
       "2                                                 accB                                                 \n",
       "3                                                 accC                                                 \n",
       "4                                                 accD                                                 \n",
       "..                                                 ...                                                 \n",
       "723                                              ycf12                                                 \n",
       "724                                               ychF                                                 \n",
       "725                                                zds                                                 \n",
       "726                                               zntA                                                 \n",
       "727                                                zwf                                                 \n",
       "\n",
       "                                               Matches Accession-1  \\\n",
       "0    octadecanal + acyl-carrier protein + NAD(P)+  ...     PMM0533   \n",
       "1    acetyl-CoA + a [carboxyl-carrier protein dimer...     PMM0534   \n",
       "2    ATP + acetyl-CoA + hydrogencarbonate  ->  ADP ...     PMM0027   \n",
       "3    a [biotin carboxyl-carrier-protein dimer]-N6-b...     PMM0060   \n",
       "4    acetyl-CoA + a [carboxyl-carrier protein dimer...     PMM0784   \n",
       "..                                                 ...         ...   \n",
       "723  RXN-15490 // oxygen[in] + 4 H+[in] + 4 e-[memb...         NaN   \n",
       "724                ATP + H2O  ->  ADP + phosphate + H+     PMM1138   \n",
       "725  beta-zeacarotene + an electron-transfer quinon...     PMM0115   \n",
       "726  a metal cation[out] + ATP + H2O  ->  a metal c...     PMM0131   \n",
       "727  D-glucopyranose 6-phosphate + NADP+  ->  6-pho...     PMM1074   \n",
       "\n",
       "      Accession-2         Gene products (polypeptides and RNAs only)  \\\n",
       "0    TX50_RS02855   long-chain acyl-[acyl-carrier-protein] reductase   \n",
       "1    TX50_RS02860  acetyl-CoA carboxylase carboxyltransferase sub...   \n",
       "2    TX50_RS00150  acetyl-CoA carboxylase biotin carboxyl carrier...   \n",
       "3    TX50_RS00320  acetyl-CoA carboxylase biotin carboxylase subunit   \n",
       "4    TX50_RS04185  acetyl-CoA carboxylase carboxyltransferase sub...   \n",
       "..            ...                                                ...   \n",
       "723           NaN       photosystem II reaction center protein Ycf12   \n",
       "724  TX50_RS06135                        redox-regulated ATPase YchF   \n",
       "725  TX50_RS00605               9,9'-di-cis-zeta-carotene desaturase   \n",
       "726  TX50_RS00690                          P-type cation transporter   \n",
       "727  TX50_RS05765                  glucose-6-phosphate dehydrogenase   \n",
       "\n",
       "                 BioCyc EC  \n",
       "0              EC:1.2.1.80  \n",
       "1              EC:2.1.3.15  \n",
       "2               EC:6.4.1.2  \n",
       "3              EC:6.3.4.14  \n",
       "4              EC:2.1.3.15  \n",
       "..                     ...  \n",
       "723  EC:1.1.3, EC:1.10.3.9  \n",
       "724            EC:3.6.1.15  \n",
       "725             EC:1.3.5.6  \n",
       "726                    NaN  \n",
       "727            EC:1.1.1.49  \n",
       "\n",
       "[728 rows x 7 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# update the EC number format to match the format in CLEAN and CLEAN-Contact's results\n",
    "biocyc_updated['BioCyc EC'] = biocyc_updated['EC-Number'].apply(format_ec)\n",
    "biocyc_updated.drop(columns=['EC-Number'], inplace=True)\n",
    "#biocyc_updated.to_csv('biocyc_updated.csv', index=False)\n",
    "biocyc_updated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c2e09ad",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:clean-contact]",
   "language": "python",
   "name": "conda-env-clean-contact-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
